src/extend.cpp

```diff
@@ -127,48 +133,143 @@ sparse_coo extend(const sparse_coo &sparse_rr_mat,
-                                         row_start_idx[i + 1] -
-                                             row_start_idx[i]);
-    std::vector<long> rr_ordered_idx = sort_indexes(rr_dists, 1);
+    long start_difference = row_start_idx[i + 1] - row_start_idx[i];
+    long n_rr_dists = std::max(0L, start_difference);
+    if (n_rr_dists > 0) {
+      Eigen::Map<Eigen::VectorXf> rr_dists(dist_vec.data() + row_start_idx[i],
+                                           row_start_idx[i + 1] -
+                                               row_start_idx[i]);
+      std::vector<long> rr_ordered_idx = sort_indexes(rr_dists, 1);
 
-    long unique_neighbors = 0;
-    float prev_value = -1;
-    for (auto rr_it = rr_ordered_idx.cbegin(); rr_it != rr_ordered_idx.cend();
-         ++rr_it) {
-      long j = std::get<1>(sparse_rr_mat)[row_start_idx[i] + *rr_it];
-      float dist = rr_dists[*rr_it];
-      if (j == i) {
-        continue;
+      long unique_neighbors = 0;
+      float prev_value = -1;
+      bool new_val;
+      for (auto rr_it = rr_ordered_idx.cbegin(); rr_it != rr_ordered_idx.cend();
+           ++rr_it) {
+        long j = std::get<1>(sparse_rr_mat)[row_start_idx[i] + *rr_it];
+        float dist = rr_dists[*rr_it];
+        if (j == i) {
+          continue;
+        }
+        if (unique_neighbors < kNN) {
+          dists[i].push_back(dist);
+          i_vec[i].push_back(i);
+          j_vec[i].push_back(j);
+          if (count_unique_distances)
+            new_val = abs(dist - prev_value) >= epsilon;
+            if (new_val)
+            {
+              unique_neighbors++;
+              prev_value = dist;
+            }
+          else
+          {
+            unique_neighbors = j_vec[i].size();
+          }
+        } else {
+          break; // next i
+        }
       }
-      bool new_val = abs(dist - prev_value) >= epsilon;
-      if (unique_neighbors < kNN || !new_val) {
-        dists.push_back(dist);
-        i_vec.push_back(i);
-        j_vec.push_back(j);
-        if (new_val) {
-          unique_neighbors++;
-          prev_value = dist;
+    }
+    len += dists[i].size();
+  }
+  // Combine the lists from each thread
+  std::vector<float> dists_all = combine_vectors(dists, len);
+  std::vector<long> i_vec_all = combine_vectors(i_vec, len);
+  std::vector<long> j_vec_all = combine_vectors(j_vec, len);
+  // Only count reciprocal matches
+  if (reciprocal_only)
+  {
+    std::vector<std::vector<float>> filtered_dists(n_samples);
+    std::vector<std::vector<long>> filtered_i_vec(n_samples);
+    std::vector<std::vector<long>> filtered_j_vec(n_samples);
+    len = 0;
+#pragma omp parallel for schedule(static) num_threads(num_threads) reduction(+:len)
+    for (long x = 0; x < i_vec_all.size(); x++)
+    {
+      if (i_vec_all[x] < j_vec_all[x])
+      {
+          for (long y = 0; y < i_vec_all.size(); y++)
+          {
+              if (i_vec_all[x] == j_vec_all[y] && j_vec_all[x] == i_vec_all[y])
+              {
```

Comment:
I have a feeling that doing this with sets may be a lot more efficient than repeated searching with a double loop. Is it ok in practise? What about with more data?

Reply:
FIne on what I've tried so far, not used very large datasets yet - if you have an alternative implementation in mind, I'm happy to compare methods!

Reply:
I think you are doing something like:
```
pairs = set()
for i, j in zip(i_vec_all, j_vec_all):
    pairs.add((i, j))
for i, j in pairs:
    if (j, i) in pairs:
       filtered.add(i, j)
```
Which is ~2N comparisons by using the set test (which is O(1)) rather than the current double loop which is N^2

Reply:
In C++ you've got `unordered_set`, and you'd probably also replace the second loop iteration with being over i_vec, j_vec again rather than over set

Reply:
Horrible to implement, mainly because it highlighted some daft bugs introduced during this PR - as of, runs a hash-based search that passes tests and gives a new output (https://microreact.org/project/giwRc2XeF8oiMMJywvFj3Q-s3reciprocalmatchesrevised) consistent with that with the previous analysis.

Reply:
Slightly better implementation in [99659b0](https://github.com/bacpop/PopPUNK/pull/232/commits/99659b03b8e22d6bbede6fbe526cdd1d028de9a9), still giving consistent outcome: https://microreact.org/project/eww6QhPCYaw27r8NoSr6JQ-s3optimisedreciprocalmatching